### ISSUE TITLE
RuntimeError: no connection to the server

### ISSUE THREAD

**Issue by alculquicondor** [mapnik/mapnik-support#29]
We are running into this errors once in a while in this call

```
mapnik.load_map(mmap, str(mapfile))
```

We have our backend server in one machine and our database in another one. They're both in the same LAN.

The error is comming from libpq. I have been checking networking issues, but it seems that there's no problem there. I made an script with an infinite loop making a query to the database (with psycopg2) each second and it doesn't fail, even though mapnik query fails at the same moment. I make a new connection for each query. It's running on the same machine as our backend server.

I don't actually know how mapnik handles connections, but could it be possible that some pool of connections is not being updated properly and it's trying to use an already closed connection?

Datasource is standard:

```
<Datasource>
  <Parameter name="type">postgis</Parameter>
  <Parameter name="host">othermachine</Parameter>
  <Parameter name="port">5444</Parameter>
  <Parameter name="user">user</Parameter>
  <Parameter name="password">password</Parameter>
  <Parameter name="dbname">dbname</Parameter>
  <Parameter name="table">tablename</Parameter>
  <Parameter name="geometry_field">"location"</Parameter>
</Datasource>
```

This is the version of mapnik we are running

```
libmapnik:     2.2.0-ubuntu1~precise2
mapnik-utils:  2.2.0-ubuntu1~precise2
python-mapnik: 2.2.0-ubuntu1~precise2

from PPA: http://ppa.launchpad.net/mapnik/v2.2.0/ubuntu/ precise/main amd64 Packages
```

PostgreSQL version: 9.3.5

**Comment by springmeyer** [comment]
Thanks for the detailed report. Here's the approach I'd recommend we take together:

1) You try Mapnik 2.3.x branch - source compiled - see if the problem persists

2) If it does, then try to reduce the testcase to something very minimal (like against a database table with no data in it). Something that you could easily set up on another machine. See if setting it up on another machine also triggers the problem. Try on a machine with postgres on the same network

3) If it still persists then pass the testcase to me to try to replicate. Again, it must be small and easy to set up otherwise I won't be able to replicate.

**Comment by alculquicondor** [comment]
Thanks for the quick answer. I'm afraid it's kind of difficult to reproduce the error, because it occurs quite randomly. We are using tilestache over mapnik and it fails just for some tiles when several requests are coming from our frontend.
We are going to try out Mapnik 2.3.x as suggested. But I think it will be a little difficult to set up an scenario. I'm going to try anyway and I'll let you know when I get something.

**Comment by springmeyer** [comment]
>  I'm afraid it's kind of difficult to reproduce the error, because it occurs quite randomly.

Unless you can help find a reliable way to reproduce then solving it will likely be impossible.

> We are using tilestache over mapnik and it fails just for some tiles when several requests are coming from our frontend.

Ah, interesting. How is tilestache deployed? mod_wsgi? gunicorn? Basically, I'm curious whether the server is multiprocess or multithreaded.

> I'm going to try anyway and I'll let you know when I get something.

Thanks.

**Comment by alculquicondor** [comment]
We're using gunicorn (with multiprocess workers, no greenlets) on top of a django project that is serving tiles from tilestache using mapnik. Is mapnik supposed to hold anything in memory?

Tested mapnik `2.3.x`. Failed for the same reason.

Things I have noticed:
- Fails less than 1% of the time
- It's more likely to fail after some time of being idle.

I'm still working on a minimal case to reproduce. I think gunicorn must be part of the stack in the scenario.

**Comment by springmeyer** [comment]
> Is mapnik supposed to hold anything in memory?

Mapnik has a per-process pool of postgres connections. It may be that if you are launching many new processing that the # of connections postgres can handle is being maxed out. You might try increasing that # (I think the default is 200).

**Comment by alculquicondor** [comment]
I have already checked that. The default is 100, but at maximum 10 are being used.

**Comment by springmeyer** [comment]
Okay, looking forward to hearing if you can reduce further.

**Comment by alculquicondor** [comment]
Found the reason of the failure: "no connection to the server" is triggered because somehow a connection (as a process) was killed or terminated, but it was still on the connection pool.

In our case, there was some nasty thing going on that eventually terminated some connections. So we fixed that. Django also had his own connection to the same server, but it didn't fail because django retries when a connection is down. I'm not sure if this should be the behavior in mapnik too.

Thank you very much for your help, it was very useful.
